modules/ocs2/src/main/scala/gem/ocs2/SequenceDecoder.scala

```diff
@@ -19,106 +19,12 @@ import Scalaz._
-    import Legacy.Instrument.Flamingos2._
-    for {
-      d <- Disperser.parse(cm)
-      e <- Legacy.Observe.ExposureTime.cparseOrElse(cm, Duration.ofMillis(0))
-      f <- Filter.parse(cm)
-      u <- Fpu.parse(cm)
-      l <- LyotWheel.parse(cm)
-      r <- ReadMode.parse(cm)
-      w <- WindowCover.cparseOrElse(cm, F2WindowCover.Close)
-    } yield DynamicConfig.F2(d, e, f, u, l, r, w)
-  }
-
-  private def gmosCommonDynamic(cm: ConfigMap): PioError \/ Gmos.GmosCommonDynamicConfig = {
-    import Legacy.Instrument.Gmos._
-
-    for {
-      x  <- XBinning.parse(cm)
-      y  <- YBinning.parse(cm)
-      ac <- AmpCount.parse(cm)
-      ag <- AmpGain.parse(cm)
-      ar <- AmpReadMode.parse(cm)
-      dx <- Dtax.parse(cm)
-      e  <- Legacy.Observe.ExposureTime.cparseOrElse(cm, Duration.ofMillis(0))
-    } yield Gmos.GmosCommonDynamicConfig(Gmos.GmosCcdReadout(x, y, ac, ag, ar), dx, e)
-  }
-
-  private def gmosCustomMask(cm: ConfigMap): PioError \/ Option[Gmos.GmosCustomMask] = {
-    import gem.ocs2.Legacy.Instrument.Gmos.{CustomMaskMdf, CustomSlitWidth}
-
-    (for {
-      f <- PioOptional(CustomMaskMdf.cparse(cm))
-      s <- PioOptional(CustomSlitWidth.parse(cm))
-    } yield Gmos.GmosCustomMask(f, s)).run
-  }
-
-  private def gmosNorthDynamic(cm: ConfigMap): PioError \/ DynamicConfig = {
-    import Legacy.Instrument.Gmos._
-    import Legacy.Instrument.GmosNorth._
-
-    val grating: PioError \/ Option[Gmos.GmosGrating[GmosNorthDisperser]] =
-      (for {
-        d <- PioOptional(Disperser.parse(cm))
-        o <- PioOptional(DisperserOrder.cparse(cm))
-        w <- PioOptional(DisperserLambda.cparse(cm))
-      } yield Gmos.GmosGrating(d, o, w)).run
-
-    for {
-      c <- gmosCommonDynamic(cm)
-      g <- grating
-      f <- Filter.parse(cm)
-      u <- Fpu.cparse(cm).map(_.flatten)
-      m <- gmosCustomMask(cm)
-      fpu = u.map(_.right[Gmos.GmosCustomMask]) orElse m.map(_.left[GmosNorthFpu])
-    } yield DynamicConfig.GmosNorth(c, g, f, fpu)
-  }
-
-  private def gmosSouthDynamic(cm: ConfigMap): PioError \/ DynamicConfig = {
-    import Legacy.Instrument.Gmos._
-    import Legacy.Instrument.GmosSouth._
-
-    val grating: PioError \/ Option[Gmos.GmosGrating[GmosSouthDisperser]] =
-      (for {
-        d <- PioOptional(Disperser.parse(cm))
-        o <- PioOptional(DisperserOrder.cparse(cm))
-        w <- PioOptional(DisperserLambda.cparse(cm))
-      } yield Gmos.GmosGrating(d, o, w)).run
-
-    for {
-      c <- gmosCommonDynamic(cm)
-      g <- grating
-      f <- Filter.parse(cm)
-      u <- Fpu.cparse(cm).map(_.flatten)
-      m <- gmosCustomMask(cm)
-      fpu = u.map(_.right[Gmos.GmosCustomMask]) orElse m.map(_.left[GmosSouthFpu])
-    } yield DynamicConfig.GmosSouth(c, g, f, fpu)
-  }
+  def decode(n: Node): PioError \/ List[Step[DynamicConfig]] =
+    (n \ "step").toList.scanLeft(EmptyConfigMap) { (m, stepNode) =>
+      stepNode.addStepConfig(m)
+    }.drop(1).traverseU(parseStep)
```

Comment:
Here I moved some code around and grouped GMOS-related functions together.  Sorry, the change looks more massive than it actually was.